staging/src/k8s.io/client-go/transport/cache.go

```diff
@@ -40,42 +43,83 @@ const idleConnsPerHost = 25
 
 var tlsCache = &tlsTransportCache{transports: make(map[tlsCacheKey]*http.Transport)}
 
+// transportIdleConnectionCollector manages a pool of TLS transports that need
+// cleanup of their idle connections. It registers one goroutine that periodically
+// invokes CloseIdleConnections on the transports in the pool, and takes them
+// out of the pool.
+type transportIdleConnectionCollector struct {
+	mu           sync.Mutex
+	transports   map[*http.Transport]struct{}
+	registerOnce sync.Once
+}
+
+// addTransport adds a TLS transport to the pool. At the first time this function
+// gets called, it registers a goroutine that runs cleanupLoop.
+func (c *transportIdleConnectionCollector) addTransport(t *http.Transport) {
+	c.registerOnce.Do(func() { go c.cleanupLoop() })
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if _, ok := c.transports[t]; !ok {
+		c.transports[t] = struct{}{}
+	}
+}
+
+// cleanupLoop does a periodical sweep of transports in the pool, calls CloseIdleConnections
+// on them, and takes them out of the pool.
+func (c *transportIdleConnectionCollector) cleanupLoop() {
+	t := time.NewTicker(defaultCleanupPeriod)
+	defer t.Stop()
+	for range t.C {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		// TODO(roycaihw): remove the log before submit
+		klog.V(1).Infof(">>> transports in the pool: %v", len(c.transports))
+		for t := range c.transports {
+			t.CloseIdleConnections()
+		}
+		c.transports = make(map[*http.Transport]struct{})
```

Comment:
if there are connections in use at the moment, and DialContext is never called again for this transport, those connections will be leaked